File: frontend/src/backupStatus.js
export function backupStatusPresentation(status) {
  if (!status) return { tone: 'pending', title: 'Проверяем бэкап', subtitle: 'Расписание 18:00 МСК' }
  if (status.state === 'completed') return { tone: 'success', title: 'Бэкап выполнен', subtitle: status.version ? `Версия ${status.version}` : 'Копия проверена' }
  if (status.state === 'failed') return { tone: 'error', title: 'Ошибка бэкапа', subtitle: 'Требуется проверка' }
  if (status.state === 'overdue') return { tone: 'error', title: 'Бэкап не выполнен', subtitle: 'Запуск в 18:00 пропущен' }
  if (status.state === 'unavailable') return { tone: 'error', title: 'Статус недоступен', subtitle: 'Проверьте службу' }
  return { tone: 'pending', title: 'Бэкап ещё не выполнен', subtitle: 'Запланирован на 18:00 МСК' }
}

export function backupStatusTooltip(status) {
  const view = backupStatusPresentation(status)
  const parts = [view.title]
  if (status?.completed_at) parts.push(`Завершён: ${new Date(status.completed_at).toLocaleString('ru-RU')}`)
  if (status?.version) parts.push(`Версия приложения: ${status.version}`)
  if (status?.database_version) parts.push(`PostgreSQL: ${status.database_version}`)
  if (status?.backup_name) parts.push(`Файл: ${status.backup_name}`)
  if (status?.size_bytes) parts.push(`Размер: ${formatBytes(status.size_bytes)}`)
  if (status?.next_run) parts.push(`Следующий запуск: ${new Date(status.next_run).toLocaleString('ru-RU')}`)
  return parts.join('\n')
}

function formatBytes(value) {
  if (value < 1024 * 1024) return `${Math.max(1, Math.round(value / 1024))} КБ`
  return `${(value / 1024 / 1024).toFixed(1)} МБ`
}

File: backend/cmd/server/backup_status.go
package main

import (
	"encoding/json"
	"errors"
	"os"
	"strings"
	"time"
)

const (
	backupScheduleHour = 18
	backupTimezone     = "Europe/Moscow"
)

type backupStatusFile struct {
	Success         bool   `json:"success"`
	CompletedAt     string `json:"completed_at"`
	BackupName      string `json:"backup_name"`
	Version         string `json:"version"`
	DatabaseVersion string `json:"database_version"`
	SizeBytes       int64  `json:"size_bytes"`
	Valid           bool   `json:"valid"`
	Error           string `json:"error,omitempty"`
}

type backupStatusResponse struct {
	State           string `json:"state"`
	Success         bool   `json:"success"`
	TodayCompleted  bool   `json:"today_completed"`
	Schedule        string `json:"schedule"`
	Timezone        string `json:"timezone"`
	CompletedAt     string `json:"completed_at,omitempty"`
	BackupName      string `json:"backup_name,omitempty"`
	Version         string `json:"version,omitempty"`
	DatabaseVersion string `json:"database_version,omitempty"`
	SizeBytes       int64  `json:"size_bytes,omitempty"`
	Valid           bool   `json:"valid"`
	NextRun         string `json:"next_run"`
	Error           string `json:"error,omitempty"`
}

func readBackupStatus(now time.Time) backupStatusResponse {
	path := getenv("BACKUP_STATUS_FILE", "/data/backup_status/latest-backup-status.json")
	raw, err := os.ReadFile(path)
	if err != nil {
		return buildBackupStatus(now, backupStatusFile{}, err)
	}
	if len(raw) > 64*1024 {
		return buildBackupStatus(now, backupStatusFile{}, errors.New("backup status file is too large"))
	}
	var value backupStatusFile
	if err = json.Unmarshal(raw, &value); err != nil {
		return buildBackupStatus(now, backupStatusFile{}, err)
	}
	return buildBackupStatus(now, value, nil)
}

func buildBackupStatus(now time.Time, value backupStatusFile, readErr error) backupStatusResponse {
	location, err := time.LoadLocation(backupTimezone)
	if err != nil {
		location = time.FixedZone("MSK", 3*60*60)
	}
	now = now.In(location)
	todayRun := time.Date(now.Year(), now.Month(), now.Day(), backupScheduleHour, 0, 0, 0, location)
	nextRun := todayRun
	if !now.Before(todayRun) {
		nextRun = todayRun.AddDate(0, 0, 1)
	}
	response := backupStatusResponse{
		State: "not_run", Schedule: "18:00", Timezone: backupTimezone,
		NextRun: nextRun.Format(time.RFC3339),
	}
	if readErr != nil {
		if !errors.Is(readErr, os.ErrNotExist) {
			response.State = "unavailable"
		}
		return response
	}
	response.Success = value.Success
	response.CompletedAt = strings.TrimSpace(value.CompletedAt)
	response.BackupName = strings.TrimSpace(value.BackupName)
	response.Version = strings.TrimSpace(value.Version)
	response.DatabaseVersion = strings.TrimSpace(value.DatabaseVersion)
	response.SizeBytes = value.SizeBytes
	response.Valid = value.Valid
	response.Error = strings.TrimSpace(value.Error)
	completed, parseErr := time.Parse(time.RFC3339, value.CompletedAt)
	if parseErr == nil {
		completed = completed.In(location)
		response.TodayCompleted = completed.Year() == now.Year() && completed.YearDay() == now.YearDay()
	}
	switch {
	case !value.Success || !value.Valid:
		response.State = "failed"
	case parseErr != nil:
		response.State = "unavailable"
	case !now.Before(todayRun) && !response.TodayCompleted:
		response.State = "overdue"
	default:
		response.State = "completed"
	}
	return response
}
